openssl-dynamic/src/main/c/jnilib.c

```diff
@@ -225,3 +154,276 @@ jint tcn_get_java_env(JNIEnv **env)
     }
     return JNI_OK;
 }
+
+// TODO: Share code with netty natives utilities.
+char* netty_internal_tcnative_util_prepend(const char* prefix, const char* str) {
+    if (prefix == NULL) {
+        char* result = (char*) malloc(sizeof(char) * (strlen(str) + 1));
+        strcpy(result, str);
+        return result;
+    }
+    char* result = (char*) malloc(sizeof(char) * (strlen(prefix) + strlen(str) + 1));
+    strcpy(result, prefix);
+    strcat(result, str);
+    return result;
+}
+
+char* netty_internal_tcnative_util_rstrstr(char* s1rbegin, const char* s1rend, const char* s2) {
+    size_t s2len = strlen(s2);
+    char *s = s1rbegin - s2len;
+
+    for (; s >= s1rend; --s) {
+        if (strncmp(s, s2, s2len) == 0) {
+            return s;
+        }
+    }
+    return NULL;
+}
+
+jint netty_internal_tcnative_util_register_natives(JNIEnv* env, const char* packagePrefix, const char* className, const JNINativeMethod* methods, jint numMethods) {
+    char* nettyClassName = netty_internal_tcnative_util_prepend(packagePrefix, className);
+    jclass nativeCls = (*env)->FindClass(env, nettyClassName);
+    free(nettyClassName);
+    nettyClassName = NULL;
+    if (nativeCls == NULL) {
+        return JNI_ERR;
+    }
+
+    return (*env)->RegisterNatives(env, nativeCls, methods, numMethods);
+}
+
+static char* netty_internal_tcnative_util_strndup(const char *s, size_t n) {
+// windows does not have strndup
+#ifdef WIN32
+    char* copy = strdup(s);
```

Comment:
can we just use `_strdup` instead?

https://msdn.microsoft.com/en-us/library/ms235454.aspx
```
These POSIX functions are deprecated. Use the ISO C++ conformant _strdup, _wcsdup, _mbsdup instead.
```